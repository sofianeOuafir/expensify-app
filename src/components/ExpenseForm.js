import React from 'react';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import moment from 'moment';

class ExpenseForm extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      description: props.expense && props.expense.description || '',
      amount: props.expense && parseFloat(props.expense.amount / 100) || '',
      note: props.expense && props.expense.note || '',
      createdAt: props.expense && moment(props.expense.createdAt) || moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }))
  };

  onDateChange = (createdAt) => { 
    if(createdAt){
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({focused}) => { this.setState(() => ({ calendarFocused: focused })) };

  onSubmit = (e) => {
    e.preventDefault();
    if(this.state.description && this.state.amount){
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
      this.setState(() => ({ error: '' }));
    } else {
      this.setState(() => ({ error: 'Please provide description and amount.' }));
    }
  };

  render () {
    return (
    <div>
      {this.state.error && <p>{this.state.error}</p>}
      <form onSubmit={this.onSubmit}>
        <input 
          type="text"
          placeholder="Description"
          autoFocus
          value={this.state.description} 
          onChange={this.onDescriptionChange} 
        />
        <input 
          type="text" 
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker 
          date={this.state.createdAt}
          onDateChange={ this.onDateChange }
          focused={ this.state.calendarFocused }
          onFocusChange={ this.onFocusChange }
          numberOfMonths={1}
          isOutsideRange={(day) => false}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <button>
          Save Expense
        </button>
      </form>
    </div>
    )
  }
}

export default ExpenseForm;
