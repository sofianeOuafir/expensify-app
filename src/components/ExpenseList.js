import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import { getVisibleExpenses } from "../selectors/expenses";

export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-item__body">
      {props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No expenses</span>
        </div>
      ) : (
        props.expenses.map((expense, index) => (
          <ExpenseListItem key={index} {...expense} />
        ))
      )}
    </div>
  </div>
);

const mapStateToProps = ({ expenses, filters }) => {
  return {
    expenses: getVisibleExpenses(expenses, filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
