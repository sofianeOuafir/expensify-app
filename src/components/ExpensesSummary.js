import React from "react";
import { connect } from "react-redux";
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';
import { getVisibleExpenses } from '../selectors/expenses'; 

export const ExpensesSummary = props => (
  <div>
    <p>Viewing {props.expenseCount} totalling {numeral(props.total / 100).format('$0,0.00')} </p>
  </div>
);

const mapStateToProps = ({ expenses, filters }) => {
  const visibleExpenses = getVisibleExpenses(expenses, filters);
  return {
    expenseCount: visibleExpenses.length,
    total: selectExpensesTotal(visibleExpenses)
  }
};

export default connect(mapStateToProps)(ExpensesSummary);
