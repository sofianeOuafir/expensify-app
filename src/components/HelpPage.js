import React from 'react';

const HelpPage = () => (
  <p>this is from my help component</p>
);

export default HelpPage;
