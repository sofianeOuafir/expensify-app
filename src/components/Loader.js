import React from 'react';

const Loader = () => (
  <div className="loader">
    <img src="/images/loader.gif" alt="loader"/>
  </div>
);

export default Loader;