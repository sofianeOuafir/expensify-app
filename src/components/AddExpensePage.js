import React from 'react';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';

class AddExpensePage extends React.Component {
  render () {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={(expense) => {
            this.props.dispatch(addExpense(expense));
            this.props.history.push('/');
          }}
        />
      </div>
    );
  }
}

export default connect()(AddExpensePage);
