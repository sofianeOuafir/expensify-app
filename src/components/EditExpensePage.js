import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  editExpense = expense => {
    this.props.editExpense(this.props.expense.id, expense).then(() => {
      this.props.history.push("/dashboard");
    });
  };

  removeExpense = () => {
    this.props.removeExpense(this.props.expense.id).then(() => {
      this.props.history.push("/dashboard");
    });
  };
  render() {
    return (
      <div>
        <h1>Edit Expense</h1>
        <ExpenseForm expense={this.props.expense} onSubmit={this.editExpense} />
        <button onClick={this.removeExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = dispatch => {
  return {
    editExpense: (id, expense) => {
      return dispatch(
        startEditExpense({
          id,
          updates: expense
        })
      );
    },
    removeExpense: id => {
      return dispatch(
        startRemoveExpense({
          id
        })
      );
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
