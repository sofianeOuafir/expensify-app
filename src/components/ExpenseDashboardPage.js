import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
import { connect } from 'react-redux';

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
