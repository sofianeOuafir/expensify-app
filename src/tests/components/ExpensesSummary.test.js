import { shallow } from 'enzyme';
import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render ExpensesSummary correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} total={3000} />);
  expect(wrapper).toMatchSnapshot();
});