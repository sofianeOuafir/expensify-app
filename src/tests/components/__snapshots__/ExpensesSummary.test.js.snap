// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly 1`] = `
<div>
  <p>
    Viewing 
    1
     totalling 
    $30.00
     
  </p>
</div>
`;
